[기송은] Week4

q45402sk opened the pull request:
## 요구사항
- [x] [기본]이메일 input에서 focus out 할 때, 값이 없을 경우 input에 빨강색 테두리와 아래에 “이메일을 입력해주세요.” 빨강색 에러 메세지가 보이나요?
- [x] [기본]이메일 input에서 focus out 할 때, 이메일 형식에 맞지 않는 값이 있는 경우 input에 빨강색 테두리와 아래에 “올바른 이메일 주소가 아닙니다.” 빨강색 에러 메세지가 보이나요?
- [x] [기본]이메일: test@codeit.com, 비밀번호: codeit101 으로 로그인 시도할 경우, “/folder” 페이지로 이동하나요?
- [x] [기본]비밀번호 input에서 focus out 할 때, 값이 없을 경우 아래에 “비밀번호를 입력해주세요.” 에러 메세지가 보이나요?
- [x] [기본]이외의 로그인 시도의 경우 이메일, 비밀번호 input에 빨강색 테두리와 각각의 input아래에 “이메일을 확인해주세요.”, “비밀번호를 확인해주세요.” 빨강색 에러 메세지가 보이나요?
- [x] [기본]이메일, 비밀번호 input에 에러 관련 디자인을 Components 영역의 에러 케이스로 적용했나요?
- [x] [기본]로그인 버튼 클릭 또는 Enter키 입력으로 로그인 되나요?

- [x] [심화]눈 모양 아이콘 클릭시 비밀번호의 문자열이 보이기도 하고, 가려지기도 하나요?
- [x] [심화]비밀번호의 문자열이 가려질 때는 눈 모양 아이콘에는 사선이 그어져있고, 비밀번호의 문자열이 보일 때는 사선이 없는 눈 모양 아이콘이 보이나요?

## 스크린샷

![스크린샷 2024-01-15 172950](https://github.com/codeit-bootcamp-frontend/4-Weekly-Mission/assets/56138199/527a2e4c-ccf0-4d77-853f-822c466be08e)

## 멘토에게

- 셀프 코드 리뷰를 통해 질문 이어가겠습니다.

q45402sk committed:
폴더 추가

q45402sk committed:
폴더 내 파일 추가

q45402sk committed:
폴더 내 파일 추가1

q45402sk committed:
classname 수정

q45402sk committed:
모바일 화면에서 이미지가 보이지 않는 현상 수정

q45402sk committed:
week4 미션

q45402sk committed:
Merge branch 'part1-기송은' of https://github.com/codeit-bootcamp-frontend/4-Weekly-Mission into part1-기송은

1005hoon commented on 위클리_미션/signin.html:
```diff
@@ -43,5 +48,6 @@
         </div>
       </form>
```
폼의 구조를 조금 나눠보면 어떨까요? 

우선 로그인 폼의 경우, 
- 이메일과 비밀번호를 입력하고, 제출하는 버튼
- 소셜 로그인을 위한 버튼 그룹
으로 크게 두 부분으로 구분이 되잖아요? 

그러면 다음과 같은 형태로 구분을 할 수 있을것 같아요. 
```
form
  form-group 
    label: 이메일
    Input: 이메일
    P: 이메일 오류 메시지
  Form-group
    label: 비밀번호
    Input: 비밀번호
    p: 비밀번호 오류 메시지
div
  p: 소셜 로그인
  button ~~~
```

1005hoon replied:
만약 이 구조로 간다면 다음과 같이 활용할 수 있겠죠 
```
<form class="singin">
    <div class="form-group">
      <label for="email">이메일</label>
      <input type="email" name="email" />
      <p class="error-message hidden" id="email-error-message">Lorem ipsum dolor sit.</p>
    </div>
    <div class="form-group">
      <label for="password">비밀번호</label>
      <input type="password" name="password" />
      <p class="error-message hidden" id="password-error-message">Lorem ipsum dolor sit.</p>
    </div>
    <div class="button-group">
      <button class="btn w-full text-white bg-primary">로그인</button>
    </div>
  </form>
  <div class="social-login">
    <h2>소셜 로그인</h2>
    <div class="flex py-3 px-2 gap-2">
      <button>카카오 로그인</button>
      <button>구글 로그인 로그인</button>
    </div>
  </div>
```

1005hoon commented on 위클리_미션/style_signin.css:
```diff
@@ -158,7 +159,62 @@ main {
+  line-height: normal;
+  position: relative;
+  top: -20px;
+}
```
위에 설명드린 마크업 구조를 따라서 한번 스타일도 개편해보시면 어떨까요?

1005hoon commented on 위클리_미션/signin.js:
```diff
@@ -0,0 +1,133 @@
+const email = document.querySelector("#email");
+const pw = document.querySelector("#pw");
+const login = document.querySelector(".login");
```
email, pw, login과 같은 변수명은 DOM 요소를 기록하기 위한 변수명으로 좋지 않아보여요! 
왜냐하면, 다른 사람이 이 변수명을 바로 본다 생각했을 때, `email` `pw` 는 실제 사용자가 입력한 그 이메일 주소와 비밀번호 값으로 생각할 수 있고, `login` 의 경우, 실제 로그인 액션을 실행시키는 함수라고 생각할 수 있기 때문이죠. 

따라서 다음과 같이 변경해보는건 어떨까요?
```
const emailInput = document.querySelector('#email-input')
const passwordInput = document.querySelector('#password-input')
const loginButton = document.querySelector('#login-button')
```

1005hoon commented on 위클리_미션/signin.js:
```diff
@@ -0,0 +1,133 @@
+const email = document.querySelector("#email");
+const pw = document.querySelector("#pw");
+const login = document.querySelector(".login");
+const regex = /^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,6}$/;
```
regex로 상수 꺼내놓는거 좋습니다!  다만 이후에 더 다양한 validation에 사용될 regex를 등록해야 할 수 있을텐데요, 이게 어떤 목적의 regex인지 써주는게 조금 더 좋을것 같아요. 

또한 조금만 더 나가본다면, 이메일 주소를 체크하는 부분에서만 이 regex가 활용될것으로 보이는데, 이메일 주소를 체크하는 validator 함수 내부에 이 regex가 존재한다면 더욱 더 관심사으 ㅣ분리 차원에서 좋을것 같습니다.

1005hoon commented on 위클리_미션/signin.js:
```diff
@@ -0,0 +1,133 @@
+const pw = document.querySelector("#pw");
+const login = document.querySelector(".login");
+const regex = /^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,6}$/;
+const eyeoff = document.querySelector(".eyeoff");
```
요녀석두 마찬가지루 dom 요소임을 표현하는 변수명으로 수정하는게 좋아보여요

1005hoon commented on 위클리_미션/signin.html:
```diff
@@ -18,11 +18,16 @@
           <button class="eyeoff"><img src="signin/eye-off.png" /></button>
         </div>
+        <div class="pw-request">비밀번호를 입력해 주세요.</div>
+        <div class="check-pw">비밀번호를 확인해 주세요.</div>
```
음, 이와 같은 오류 메시지들은 우선 P 태그로 두는게 좋아보입니다. 
그리고 JavaScript는 DOM 요소를 조작할 수 있죠? 그 뜻은, p 태그의 클래스 목록뿐만 아니라 p 태그가 포함하고 있어야 할 콘텐츠까지 조작이 가능하다는 뜻인데요. 
그런만큼 오류 메시지의 콘텐츠들과 보여지고 숨겨지는 로직은 자바스크립트로 구현해 동적으로 동작하도록 해주고, 마크업 단계에서는 위치와 에러메시지들이 기본적으로 지녀야 할 스타일링만 정해두면 어떨까요? 

```
<p id="email-error-message" class="text-xs text-red font-medium hidden"></p>

const errorMessage = document.querySelector('#email-error-message');
errorMessage.classList.push("block");
errorMessage.textContent = "오류 메시지"
```

1005hoon commented on 위클리_미션/signin.js:
```diff
@@ -0,0 +1,133 @@
+eyeoff.addEventListener("click", function (event) {
+  event.preventDefault();
+  pwVisibility();
+});
```
함수를 조금 더 모듈러 하게 작성해봐야 할것 같아요! 아직 시간 있으니 다음과 같은 형태로 구조를 변경해보는건 어떨까요? 

```
/**
 * HTML DOM Elements
 */
const emailInput = document.querySelector('#email-input')
const passwordInput = document.querySelector('#password-input')
const loginButton = document.querySelector('#login-button')
const showPasswordButton = document.querySelector(".eyeoff");
/**
 * UI Event Listeners
 */

/**
 * Utility Logics
 */
function isEmailFormat(emailString) {
  const regex = /^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,6}$/;
  return regex.test(emailString)
}

function isEmpty(text) {
  return text.length === 0
}

/**
 * Validates a user's login information.
 *
 * This function checks if the provided login information matches a predefined valid user schema.
 * It compares both email and password from the loginInfo object with the valid user credentials.
 *
 * @param {Object} loginInfo - An object containing the user's login information.
 * @param {string} loginInfo.email - The email address of the user.
 * @param {string} loginInfo.password - The password of the user.
 * @returns {boolean} - Returns `true` if the email and password match the valid user schema, otherwise returns `false`.
 */
function isValidUser(loginInfo) {
  const validUserSchema = { email: "codeit@codeit.com", password: "password" }

  const { email, password } = loginInfo
  
  if (email !== validUserSchema.email) {
    return false
  }

  if (password !== validUserSchema.password) {
    return false
  }

  return true
}

function validateUserLoginInfo() {
  const email = emailInput.value; 

  const isEmailEmpty = isEmpty(email)
  if (isEmailEmpty) {
    toggleError(emailErrorMessage, "이메일을 입력하세요")
    return false
  }

  const isValidEmailFormat = isEmailFormat(email)
  if (!isValidEmailFormat) {
    toggleError(emailErrorMessage, '잘못된 이메일 유형입니다')
    return false
  }

  const password = passwordInput.value;
  const isPasswordEmpty = isEmpty(password)
  if (isPasswordEmpty) {
    toggleError(passwordErrorMessage, "비밀번호를 입력하세요")
    return false
  }

  return true
}
/**
 * Service Logic
 */
function handleLogin(e) {
  const isValidInput = validateUserLoginInfo()
  if (!isValidInput) {
    return false
  }

  // login 로직 실행
  const user = isValidUser({ email, password });
  
  if (!user) {
    toggleError(emailErrorMessage, "이메일을 확인하세요")
    toggleError(passwordErrorMessage, "비밀번호를 확인하세요")
    return
  }

  location.href = "로그인 성공 후 이동 될 엔드포인트"
}
```

1005hoon commented on 위클리_미션/signin.js:
```diff
@@ -0,0 +1,133 @@
+    pwrqs.classList.remove("pw-request-show");
+  }
+}
+
```
사용자의 액션을 기준으로 함수를 작성하는것 보단, 실행되어야 하는 로직의 단위를 구성해서 함수를 작성하는게 조금 더 객체지향적이고 관리가 유용할 것 같아요! 

지금 작성된 함수들은 사용자가 특정 액션을 취했을 때 발생해야 하는 모든 액션들을 절차지향적으로 작성하셨는데요, 이 경우, 특정 로직이 어느단계에서 발생했는지 관리하기가 어려워지게 됩니다. 따라서 조금 더 객체지향적으로 나누어서 함수 하나가 딱 한 단위의 작업, 책임 그리고 역할을 갖도록 하는게 좋아보여요.

이에 대한 예시로 아래에 작성했으니 한번 참고해보시구 이해가 잘 안되는 부분이 있다면 바로 디스코드로 콜 주세요!

1005hoon reviewed (changes requested):
전반적으로 폴더 구조를 한번 정리하는게 필요할것 같아 보입니다! 
우선 지금 리포지토리에서 프로젝트라는 폴더 안에 모든 코드파일들이 담겨있는데요, 프로젝트라는 폴더로 두기보단, repository로 설정된 폴더를 진입로로 활용하는게 더 좋아보여요. 
그 상태에서, 페이지별로 (홈페이지, 로그인페이지 등등) 폴더를 만들어 각 페이지에서 활용될 파일들 (index.html, style.css, index.js) 파일을 만들어 활용하는게 이후 코드 파일 유지보수에도 더 도움이 될 것 같아요!. 

그 외적으로는, UI 콤포넌트 구조를 작성할 때 조금 더 레이아웃 관점에서 어떻게 블록블록을 묶으면 좋을지 구상을 해보면 좋을것 같습니다! 이에 대해서는 코드단의 리뷰에서 디테일하게 작성해놓았기 때문에 한번 참고해보시면 좋을것 같구요. 

자바스크립트의 경우 생각보다 잘 작성해주셔서 놀랐네요 ㅎㅎ 더 개선시킬 수 있는 방향, 역시 코드단의 리뷰로 남겨두었으니 한번 확인 부탁드릴게요!

q45402sk committed:
refactor: 공통 레이아웃 및 UI 스타일링 개편

3주차 코드리뷰 반영:
1. 폴더 구조와 파일 네이밍
2. 공통 모듈화
3. 반응형을 우선한 프로젝트 구조

q45402sk committed:
refactor: signin/index.js 수정

4주차 코드리뷰 반영
-signin/index1.js: 함수를 모듈러하게 작성
-signin/index.html: 폼의 구조 변경
-signin/index.js: 변수명 변경

1005hoon commented on hompage.css:
```diff
@@ -239,7 +234,7 @@ p {
 
 .screenshot {
-  background-image: url("landing_image/_img4\ 1.png");
+  background-image: url("public/images/landing_image/_img4-1.png");
```
좋습니다! 조금 더 나아가서 `_img4-1.png` 를 이 이미지가 표현되는 이미지 제목으로 붙여봐도 좋을것 같아요. 만약 이 이미지에 대한 특징을 정리하기 어렵다면, 이 이미지가 사용되는 구간을 이미지 이름으로 작성해보시는것도 좋아요.
`public/images/feature-4.png` 처럼요!

1005hoon commented on signin/index.html:
```diff
@@ -0,0 +1,82 @@
+            <!-- 비밀번호를 입력해 주세요. -->
+            <p class="pw-request"></p>
+            <!-- 비밀번호를 확인해 주세요. -->
+            <p class="check-pw"></p>
```
error message component 는 이메일과 비밀번호 각각 하나만 존재해도 괜찮을것 같아요.
각 component에 들어갈 메시지 콘텐츠는 자바스크립트로 동적으로 할당하면 되니깐요!

1005hoon commented on signin/index.js:
```diff
@@ -0,0 +1,141 @@
+    event.preventDefault();
+    loginFunction();
+  }
+});
```
form 내부에 버튼이 클릭 또는 엔터로 인해 트리거 될 때 발생하는 `formSubmitAction` 을 활용해보는 방법으로 개편할 수 있을것 같아요!

1005hoon commented on signin/index.js:
```diff
@@ -0,0 +1,141 @@
+    pwRequest();
+    emailChecker();
+  }
+}
```
함수 이름에 `function` 이라는 단어는 최대한 사용하지 않는게 좋습니다! 
그리구 로직의 경우, 보통은 validation이 먼저 실행되고 서비스로직이 그 다음에 호출되는 순서로 진행됩니다! 
이 형태를 early fail이라고 하는데요, 멘토링 세션에서 한번 이야기를 해보도록 할게요

1005hoon commented on signin/index.js:
```diff
@@ -0,0 +1,141 @@
+  if (pwrqs) {
+    pwrqs.classList.remove("pw-request-show");
+  }
+}
```
변하지 않는 변수들에 대해서는 let보다는 const를 사용하는게 좋을것 같습니다! 
querySelector도 가져온 DOM element들의 경우, 지금 변경하는 요소들이 객체 내부의 값들을 바꾸는거여서 굳이 let을 활용하지 않아도 괜찮아요!

1005hoon commented on signin/index1.js:
```diff
@@ -0,0 +1,164 @@
+showPasswordButton.addEventListener("click", function (event) {
+  event.preventDefault();
+  pwMasking();
+});
```
아 여기에 업데이트된게 있었군요! 훨씬 좋습니다 ㅋㅋ

1005hoon reviewed (approved):
뒤에 추가해주신걸 나중에 봐서 기존 코드 리뷰를 조금 더 추가해버리게 되었네요 ㅎㅎ
모듈러 하게 작성하신부분 너무 좋구요, 혹시나 이해 안되는 부분 있으시면 너무 혼자 고민하지 마시고 바로 디스코드에 남겨주세요! 
이번 4주차 작업은 여기서 클로징 하도록 하겠습니다. 고생하셨습니다
